src/runtime_src/core/edge/sw_em/generic_pcie_hal2/halapi.cxx

```diff
@@ -76,26 +76,31 @@ xclDeviceHandle xclOpen(unsigned deviceIndex, const char *logfileName, xclVerbos
   FeatureRomHeader fRomHeader;
   std::memset(&fRomHeader, 0, sizeof(FeatureRomHeader));
 
-  xclcpuemhal2::CpuemShim *handle = nullptr;
+  xclcpuemhal2::CpuemShim *shimObj = NULL;
+  void* handle;
+
   bool bDefaultDevice = false;
   std::map<unsigned int, xclcpuemhal2::CpuemShim*>::iterator it = xclcpuemhal2::devices.find(deviceIndex);
   if(it != xclcpuemhal2::devices.end())
   {
-    handle = (*it).second;
+    xclcpuemhal2::shimHandleMap[handle] = (*it).second;
   }
   else
-  {
-    handle = new xclcpuemhal2::CpuemShim(deviceIndex,info,DDRBankList,false,false,fRomHeader);
+  {    
+    shimObj  = new xclcpuemhal2::CpuemShim(deviceIndex, info, DDRBankList, false, false, fRomHeader);;
+    xclcpuemhal2::shimHandleMap[handle] = shimObj;
     bDefaultDevice = true;
-    xclcpuemhal2::devices[deviceIndex++] = handle;
+    xclcpuemhal2::devices[deviceIndex++] = shimObj;
   }
 
   if (!xclcpuemhal2::CpuemShim::handleCheck(handle)) {
-    delete handle;
-    handle = 0;
+    delete shimObj;
+    shimObj = 0;
```

Comment:
You need to remove shimObj entry from the map aswell. Stale shim objects would be there in map otherwise.